Make quiesce timeout configurable on server element

jimblye opened the pull request:
- [x] I have considered the risk of behavior change or other zero migration impact 
- [x] Fixes #25523
- [x]  Fixes #25523
- [Link to UFO](https://ibm.ent.box.com/integrations/officeonline/openOfficeOnline?fileId=1336135718379&sharedAccessCode=)

#######################################################

The new attribute is configured as follows:
```
<server description="Production Server" quiesceTimeout="60s">
    <!-- Other server configuration -->
</server>
```


It is best reviewed one commit at a time.  
### Commits:
1.  Remove beta code - quiesceTimeout attr on executor element

2. Adds storage for quiesceTimeout and an SPI to access it.  The \<server\> "description" attribute of the server element was already stored in BaseConfiguration.  So "quiesceTimeout" is added there as well.  The value is ultimately needed in the RuntimeUpdateManagerImpl.  Unfortunately, com.ibm.ws.config already depends on com.ibm.ws.runtime.  To avoid a circular dependency the ServerElementConfig service is created.  The implementation is in com.ibm.ws.config, which delegates to BaseConfiguration.  The interface is in com.ibm.ws.kernel.service.

3. Update com.ibm.ws.kernel.service bundle version

4. Adds the new metatype for "server" and adds the parsing for the quiesceTimeout in the XMLConfigParser.  Parsing of the "description" attribute of the "server" element was already in XMLConfigParser.    The default timeout comes from the metatype registry.

5. Adds an OSGi service reference to ServerElementConfig in RuntimeUpdateManagerImpl. Update quiesce() method to get timeout from service. Modify ExecutorServiceImpl.quiesceThreads() to receive timeout as parameter instead of reading from configuration. Update ThreadQuiesce interface.

6. Tests: units & FATs.  I was able to re-use tests from old `<executor>` element implementation with slight modifications.

7. Update  SchemaWriter.java (schemaGen) to write <server> attributes 

8. Minor refactoring of SchemaWriter


### Basic flow 
[UFO link](https://ibm.ent.box.com/file/1336135718379?s=28jbu911otywabyrxg9r1r8c76dzv0dt) has larger picture
![quiesce_architecture](https://github.com/user-attachments/assets/0f8e7953-04cd-452f-a8b8-669d4f1dba2f)

### TO DO
- [x]  Build is good 
- [x]  Verify schemaGen & serverSchemaGen both show the new attribute.
- [x]  Verify that quiesceTimeout can be dynamically updated (i.e. after server start).
- [x]  Check if documentation for quiesceTimeout is automatically generated
- [x]  Update to schemagenWriter should be beta for now
- [x]  Are there any hits to instant on? @tjwatson

jimblye commented:
#build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=9a9b446c-b051-43a9-89cb-bd0c6ea3f729) - :hourglass_flowing_sand:InProgress)

_Note: Target locations of links might be accessible only to IBM employees._

jimblye commented:
#build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=6ac26c0e-045f-49af-a805-73e5cc6724b1) - :hourglass_flowing_sand:InProgress)
#libby

_Note: Target locations of links might be accessible only to IBM employees._

jimblye commented:
#build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=93c83970-49d1-437e-a8f0-54dd54374996) - :hourglass_flowing_sand:InProgress)
#libby

_Note: Target locations of links might be accessible only to IBM employees._

jimblye commented:
#build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=afa01b18-3add-4d47-95f7-c59920aeb32c) - :hourglass_flowing_sand:InProgress)
#libby

_Note: Target locations of links might be accessible only to IBM employees._

jimblye commented:
#build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=26c26ba8-a3f6-4b25-9821-79e711aeb7e4) - :hourglass_flowing_sand:InProgress)
#libby

_Note: Target locations of links might be accessible only to IBM employees._

jimblye commented:
#build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=000f05b4-989b-40b0-9a7d-8c4f3d8a54c7) - :hourglass_flowing_sand:InProgress)
#libby

_Note: Target locations of links might be accessible only to IBM employees._

jantley-ibm commented on dev/com.ibm.ws.kernel.server.element/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -0,0 +1,26 @@
+server.description.desc=Description of the server.
+
+server.quiesce.timeout.name=Quiesce Timeout
+server.quiesce.timeout.desc=The maximum time to wait for the server to quiesce during shutdown. Specify a positive integer followed by a unit of time: hours (h), minutes (m), or seconds (s). If a unit is omitted after a positive integer, seconds are assumed. Values can be combined (for example, 1m30s is equivalent to 90s and also 90). The default value is 30s.
```
change
The maximum time to wait
to
The maximum length of time to wait

jimblye replied:
done.

jantley-ibm commented on dev/com.ibm.ws.threading/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -26,7 +26,4 @@ core.threads=Core threads
-quiesce.timeout=Quiesce timeout
-quiesce.timeout.desc=The quiesce timeout is the amount of time the server allows services to perform pre-shutdown work. For example, inbound listeners are stopped, but existing requests are allowed to complete.
\ No newline at end of file
+max.threads.desc=Maximum number of threads that can be associated with the executor. If greater than 0, maxThreads can be a minimum of 4, and should be at least as large as coreThreads; if maxThreads is set less than coreThreads, Liberty will reduce coreThreads to the maxThreads value.  If the value of maxThreads is less than or equal to 0, the maximum number of threads is unbounded, which lets t ...
```
change
Maximum number of threads
to
The maximum number of threads

change
Liberty will reduce
to
Liberty reduces

change
unbounded, which lets the Liberty kernel decide how many threads to associate with the executor without having a defined upper boundary.
to
unbounded; in this case, the Liberty kernel sets the number of threads to associate with the executor.

jimblye replied:
That particular message isn't supposed to change.  It's an old message.   All I did to dev/com.ibm.ws.threading/resources/OSGI-INF/l10n/metatype.properties was to remove quiesce.timeout message.

I don't know why "max.threads.desc" message shows up as a change.  It looks the same as the old message .

jimblye replied:
I think "max.threads.desc" shows up as a difference because I removed a carriage return.  Since the message isn't new, I won't change it.  I'll just add the carriage return.

jantley-ibm reviewed (commented):
Just a few review comments from ID. Let me know if you have any questions or concerns about suggested changes.

jimblye commented:
#build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=09136aec-a3ee-4e03-8e2b-f1bb0e6b4571) - :hourglass_flowing_sand:InProgress)
#libby

_Note: Target locations of links might be accessible only to IBM employees._

jimblye commented:
#build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=68cf8687-0701-4d07-aaa2-9145050f8a2b) - :hourglass_flowing_sand:InProgress)
#libby

_Note: Target locations of links might be accessible only to IBM employees._

jimblye commented:
### schemaGen output
<img width="826" height="701" alt="image" src="https://github.com/user-attachments/assets/d4ccd1e6-9152-4649-9963-5310d89f9743" />

jimblye commented:
#build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=ffb4989a-46b4-4a97-b8a9-4da3c9f184e4) - :hourglass_flowing_sand:InProgress)
#libby

_Note: Target locations of links might be accessible only to IBM employees._

jimblye commented:
#build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=e19b36c4-45b3-4e1e-8516-f729b130464c) - :hourglass_flowing_sand:InProgress)

_Note: Target locations of links might be accessible only to IBM employees._

jimblye commented:
#build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=77e27afc-a4bd-4492-9653-25b8a716a3ec) - :hourglass_flowing_sand:InProgress)

_Note: Target locations of links might be accessible only to IBM employees._

jimblye committed:
Remove beta code -quiesceTimeout attr on executor element

jimblye committed:
Remove beta code -quiesceTimeout attr on executor element

jimblye committed:
Add API and storage for quiesceTimeout attribute

jimblye committed:
Add API and storage for quiesceTimeout attribute

jimblye committed:
Update com.ibm.ws.kernel.service bundle version

jimblye committed:
Update com.ibm.ws.kernel.service bundle version

jimblye committed:
Add Server Element Metatype and server.xml Parsing

jimblye committed:
Add Server Element Metatype and server.xml Parsing

jimblye committed:
Update RuntimeUpdateManagerImpl and ExecutorServiceImpl

jimblye committed:
Update RuntimeUpdateManagerImpl and ExecutorServiceImpl

jimblye committed:
Add tests for quiesceTimeout

jimblye committed:
Add tests for quiesceTimeout

jimblye committed:
Update SchemaWriter to handle attributes on the server element

jimblye committed:
Update SchemaWriter to handle attributes on the server element

jimblye committed:
minor refactor SchemaWriter

jimblye committed:
minor refactor SchemaWriter

SchemaWriter - respect beta flag - fix detection of EARLY_ACCESS

jimblye committed:
SchemaWriter - respect beta flag - fix detection of EARLY_ACCESS

jimblye committed:
bump bundle version of com.ibm.websphere.appserver.api.kernel.service

jimblye committed:
bump bundle version of com.ibm.websphere.appserver.api.kernel.service

Update kernel-1.0.mf for API bundle version 1.2

jimblye committed:
Update kernel-1.0.mf for API bundle version 1.2

jimblye committed:
Change timeout to milliseconds instead of seconds

jimblye committed:
Change timeout to milliseconds instead of seconds

jimblye committed:
Add dynamic update test for quiesceTimeout

jimblye committed:
Add dynamic update test for quiesceTimeout

jimblye committed:
Update XMLConfigParserTest.java to use getQuiesceTimeoutMillis API

NoClassDefFoundError - com.ibm.wsspi.kernel.service.utils.PathUtils

Add Require-Bundle to com.ibm.ws.kernel.feature bnd.bnd

jimblye committed:
Update XMLConfigParserTest.java to use getQuiesceTimeoutMillis API

jimblye committed:
NoClassDefFoundError - com.ibm.wsspi.kernel.service.utils.PathUtils

jimblye committed:
Add Require-Bundle to com.ibm.ws.kernel.feature bnd.bnd

jimblye committed:
Message update and minor metatype change

jimblye committed:
Message update and minor metatype change

jimblye commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=a757cc28-0e4a-4444-a6ca-70530b88bbfb) - :hourglass_flowing_sand:InProgress)

_Note: Target locations of links might be accessible only to IBM employees._

jimblye commented:
!build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=af6027f0-750e-4679-b031-04a288e05470) - :hourglass_flowing_sand:InProgress)
git.clone.checkout.branch=25523-Make-quiesce-time-configurable-server-element-with-DS
git.clone.checkout.commit=9b77b057f435fbd7adf4c78cd4793846614eaa98
git.clone.repository.sshUrl=git@github.ibm.com:jimblye/WS-CD-Open.git
git.clone.repository.username=jimblye

_Note: Target locations of links might be accessible only to IBM employees._

isaacrivriv commented on dev/io.openliberty.channel.config.common/src/io/openliberty/channel/config/internal/ChannelFrameworkConfigImpl.java:
```diff
@@ -206,6 +268,11 @@ private void setDefaultChainQuiesceTimeout(Object value) {
+        
+        // Log the final value being used for testing purposes
+        if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+            Tr.event(tc, "Final chainQuiesceTimeout value: " + this.chainQuiesceTimeout + "ms");
```
This should be more a debug message IMO

jimblye replied:
Changed to Tr.debug

isaacrivriv commented on dev/com.ibm.ws.config_fat/publish/files/quiesceoverride/chainQuiesceTimeoutOnly.xml:
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Should we add copyrights to these xmls? I know for most of the http testing, the server.xmls have copyrights.

jimblye replied:
About half of the server.xml files in the code base have copyrights, and the other half don't.  Here's a snippet of what Bob had to say:
### Copyright Law Perspective

**Configuration files typically don't meet the threshold for copyright protection:**

1. **Lack of Originality**: Simple feature lists and standard configurations lack sufficient originality
2. **Functional Nature**: They serve a purely functional purpose (configuring software)
3. **Merger Doctrine**: The expression (XML configuration) is inseparable from the idea (how to configure Liberty)
4. **Scenes à Faire**: Standard configurations are expected/necessary ways to accomplish tasks

### Industry Practice

**Most open-source projects don't copyright configuration files:**
- Apache projects: Configuration files typically lack copyright headers
- Spring Boot: application.properties/yml files have no copyright
- Kubernetes: YAML configuration files have no copyright
- Docker: Dockerfiles typically have no copyright

## Recommendation

### Do NOT Require Copyright Statements on server.xml Files

isaacrivriv commented on dev/io.openliberty.channel.config.common/resources/io/openliberty/channel/config/internal/resources/ChannelConfigMessages.nlsprops:
```diff
@@ -0,0 +1,33 @@
+###############################################################################
+# Copyright (c) 2026 IBM Corporation and others.
```
I was thinking we can reuse the resources [here](https://github.com/OpenLiberty/open-liberty/blob/0379d77387604661985100b1303bc1e0145d49b8/dev/com.ibm.ws.channelfw/resources/com/ibm/ws/channelfw/internal/resources/ChannelfwMessages_pl.nlsprops) but thinking about this more, I don't think the config bundle and the channelfw bundle should have dependencies on each other so this makes sense.

isaacrivriv replied:
No action to do here, just commented for future.

jimblye replied:
No action taken

isaacrivriv reviewed (commented):
Simple comments for the chain quiesce changes. I think we should also add the `Co-authored-by-AI` reference for the commits that were done with Bob since I didn't see anything like that.

jimblye committed:
When configured - quiesceTimeout overrides chainQuiesceTimeout

jimblye commented:
!build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=94869c2b-a855-45b5-83b0-6d8d24bbc0db) - :warning: completed with warnings.)
git.clone.checkout.branch=25523-Make-quiesce-time-configurable-server-element-with-DS
git.clone.checkout.commit=9b77b057f435fbd7adf4c78cd4793846614eaa98
git.clone.repository.sshUrl=git@github.ibm.com:jimblye/WS-CD-Open.git
git.clone.repository.username=jimblye

_Note: Target locations of links might be accessible only to IBM employees._

jimblye commented:
!build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=55e2fc01-cd42-49e9-adfc-2a1073a75b99) - :x: completed with errors/failures.)
git.clone.checkout.branch=25523-Make-quiesce-time-configurable-server-element-with-DS
git.clone.checkout.commit=9b77b057f435fbd7adf4c78cd4793846614eaa98
git.clone.repository.sshUrl=git@github.ibm.com:jimblye/WS-CD-Open.git
git.clone.repository.username=jimblye

_Note: Target locations of links might be accessible only to IBM employees._

rsherget commented on dev/com.ibm.ws.config/src/com/ibm/ws/config/schemagen/internal/SchemaWriter.java:
```diff
@@ -167,7 +173,7 @@ private static File getInstallDir() {
                         File f = new File(url.toURI());
                         // The parent of the jar is lib, so the parent of the parent is the install.
-                        installDir = f.getParentFile();
+                        installDir = f.getParentFile().getParentFile();
```
Did this not get the installDir before?

rsherget reviewed (approved):
LGTM pending build

jimblye commented:
!build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=1d2e97a6-e5ee-4d84-bf2c-26d5e468701f) - :warning: completed with warnings.)
git.clone.checkout.branch=25523-Make-quiesce-time-configurable-server-element-with-DS
git.clone.checkout.commit=9b77b057f435fbd7adf4c78cd4793846614eaa98
git.clone.repository.sshUrl=git@github.ibm.com:jimblye/WS-CD-Open.git
git.clone.repository.username=jimblye

_Note: Target locations of links might be accessible only to IBM employees._

Azquelt commented:
This was reverted in #35475 because it broke the continuous build.
